File: typescript/ch1/showStudent.ts
import { StudentDB } from './db';
import fp from 'lodash/fp';

const studentDB = new StudentDB();
const ssn = '444-44-4444';

const showStudent = [
  () => {
    const student = studentDB.find(ssn);
    if (student != null) {
      console.log('showStudent_ip', `${student.ssn}, ${student.firstname}, ${student.lastname}`);
    } else {
      throw new Error('학생을 찾을 수 없습니다.');
    }
  },
  
  () => {
    const findFrom = fp.curry((db: any, id: string) => {
      const obj = db.find(id);
      if (obj == null) {
        throw new Error('학생을 찾을 수 없습니다.');
      }
  
      return obj;
    });
  
    const makeString = (student: any) => `${student.ssn}, ${student.firstname}, ${student.lastname}`;
    const log = (info: string) => console.log('showStudent_fp', info);
  
    const showStudent = fp.compose(
      log,
      makeString,
      findFrom(studentDB)
    );
  
    showStudent(ssn);
  }
];

export default showStudent;

File: python/python/ch1/show_student.py
from .db import StudentDB
import toolz as tz

student_db = StudentDB()
ssn = '444-44-4444'

def show_student_ip():
    student = student_db.find(ssn)
    if student != None:
        print('show_student_ip', f'{student["ssn"]}, {student["firstname"]}, {student["lastname"]}')
    else:
        raise ValueError('학생을 찾을 수 없습니다.')
      
def show_student_fp():
    def db_find(db, id):
        obj = db.find(id)
        if obj == None:
            raise ValueError('학생을 찾을 수 없습니다.')
        return obj
      
    find_from = tz.curry(db_find)
    make_string = lambda student: f'{student["ssn"]}, {student["firstname"]}, {student["lastname"]}'
    log = lambda info: print('show_student_fp', info)
    
    show_student = tz.compose(
      log,
      make_string,
      find_from(student_db)
    )
    
    show_student(ssn)